examples/allocation/zig/testdata/greet.zig

```diff
@@ -0,0 +1,79 @@
+const std = @import("std");
+const allocator = std.heap.page_allocator;
+
+// main is required for Zig to compile to enable WASI.
+pub fn main() anyerror!void {
+    return;
+}
+
+extern "env" fn log(ptr: u32, size: u32) void;
+
+// _log prints a message to the console using log.
+pub fn _log(message: []u8) void {
+    const ptrSize = stringToPtr(message);
+    var p = ptrSize >> 32;
+    log(@truncate(u32, p), @truncate(u32, ptrSize));
+}
+
+pub export fn malloc(size: usize) usize {
+    var memory = allocator.alloc(u8, size) catch unreachable;
+    return @ptrToInt(memory.ptr);
+}
+
+pub export fn free(ptr: usize) void {
+    var s: []u8 = undefined;
+    s.ptr = @intToPtr([*]u8, ptr);
+    allocator.free(s);
+}
+
+pub fn _greeting(name: []u8) ![]u8 {
+    return try std.fmt.allocPrint(
+        allocator,
+        "Hello, {s}!",
+        .{name},
+    );
+}
+
+// _greet prints a greeting to the console.
+pub fn _greet(name: []u8) void {
+    const g = _greeting(name);
+    const s = std.fmt.allocPrint(
+        allocator,
+        "wasm >> {s}",
+        .{g},
+    ) catch unreachable;
+    _log(s);
+}
+
+// greet is a WebAssembly export that accepts a string pointer (linear memory offset) and calls greet.
+pub export fn greet(ptr: u32, size: u32) void {
+    const name = ptrToString(ptr, size);
+    _greet(name);
+}
+
+// greeting is a WebAssembly export that accepts a string pointer (linear memory
+// offset) and returns a pointer/size pair packed into a uint64.
+//
+// Note: This uses a uint64 instead of two result values for compatibility with
+// WebAssembly 1.0.
+pub export fn greeting(ptr: u32, size: u32) u64 {
+    const name = ptrToString(ptr, size);
+    const g = _greeting(name) catch unreachable;
+    return stringToPtr(g);
+}
+
+// ptrToString returns a string from WebAssembly compatible numeric types
+// representing its pointer and length.
+pub fn ptrToString(ptr: u32, size: u32) []u8 {
```

Comment:
This function shouldn't be needed as rather than accepting `u32` functions such as `greeting` can accept a many-item pointer (`[*]u8`) and can then be sliced to create a slice like as follows:
```zig
pub export fn greeting(message: [*]const u8, size: usize) u64 {
    const g = _greeting(messages[0..size]) catch return 0;
    return stringToPtr(g);
}
```

Reply:
Brilliant. Fixed.
I couldn't use `[*]const u8` because `const` doesn't allow slicing, maybe? I'm new to Zig so please let me know if I'm wrong.
https://github.com/tetratelabs/wazero/pull/741/commits/2c904c3c1698a8ef53edcf523a159a90244f2bab

Reply:
> Brilliant. Fixed. I couldn't use `[*]const u8` because `const` doesn't allow slicing, maybe? I'm new to Zig so please let me know if I'm wrong. [9c349ee](https://github.com/tetratelabs/wazero/commit/9c349eef38afa1b6b6e6f508681f4539c20f4ba6)

It's because the function signature in `_greeting` accepts an `[*]u8` so to accept `[*]const u8` you'd have to update that also.

Reply:
It makes sense. Thanks.
https://github.com/tetratelabs/wazero/pull/741/commits/d964f1f9a23eb0d4e8ad6e8c1e9ccfc3acad048b